File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;

class Solution {
    static String[] words = {"A", "E", "I", "O", "U"};
    static List<String> list;
        
    public int solution(String word) {
        int answer = 0;
        list = new ArrayList<>();
        dfs("", 0);
        
        for(int i = 0; i < list.size(); i++){
            if(list.get(i).equals(word)){
                answer = i;
                break;
            }
        }
        
        return answer;
    }
    
    static public void dfs(String str, int len){
        list.add(str);
        if(len == 5){return;}
        
        for(int i = 0; i < 5; i++){
            dfs(str+words[i], len+1);
        }
    }
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
def solution(word):
    answer = 0
    alphabet = ['A', 'E', 'I', 'O', 'U']
    found = False
    
    def make_word(w):
        nonlocal answer, found
        
        for ch in alphabet:
            nw = w + ch
            answer+=1
            
            if nw == word:
                found = True
                return
            if len(nw) < 5:
                make_word(nw)
                if found:
                    return
            
    make_word('')
    
    
    return answer


# AAAAA (5) > AAAAE(6) > ... > E > EE > EEE > EEEE
